### ISSUE TITLE
cp.roll cannot take array as shift argument

### ISSUE THREAD

**Issue by thomasaarholt** [cupy/cupy#4476]
I was taking the output of cp.unravel_index, which returns an array, and feeding it to cp.roll, which shifts an array. This raises the error below, but works when the shift argument is a tuple. It applies to 1D shifts as well.

It looks like perhaps the `numpy.broadcast(shift, axis)` in the error below should be `cupy.broadcast(shift, axis)` instead?

```python
>>> arr = cp.array([1,2,3,4]).reshape((2,2))
>>> shift = cp.array([1,1]) # shift (1,1) works
>>> cp.roll(arr, shift)

---------------------------------------------------------------------------
TypeError                                 Traceback (most recent call last)
<ipython-input-158-005527abcbc1> in <module>
      1 arr = cp.array([1,2,3,4])
      2 shift = cp.array([1])
----> 3 cp.roll(arr, shift)

~\Miniconda3\envs\py\lib\site-packages\cupy\_manipulation\rearrange.py in roll(a, shift, axis)
     99     """
    100     if axis is None:
--> 101         return roll(a.ravel(), shift, 0).reshape(a.shape)
    102     else:
    103         axis = _reduction._get_axis(axis, a.ndim)[0]

~\Miniconda3\envs\py\lib\site-packages\cupy\_manipulation\rearrange.py in roll(a, shift, axis)
    103         axis = _reduction._get_axis(axis, a.ndim)[0]
    104 
--> 105         broadcasted = numpy.broadcast(shift, axis)
    106         if broadcasted.nd > 1:
    107             raise ValueError(

cupy\core\core.pyx in cupy.core.core.ndarray.__array__()

TypeError: Implicit conversion to a NumPy array is not allowed. Please use `.get()` to construct a NumPy array explicitly.
```

CuPy Version                 : 8.2.0
NumPy Version                : 1.19.4
SciPy Version                : 1.5.3
Cython Build Version         : 0.29.21
CUDA Root                    : C:\Program Files\NVIDIA GPU Computing Toolkit\CUDA\v11.1
CUDA Build Version           : 11010
CUDA Driver Version          : 11020
CUDA Runtime Version         : 11010
cuBLAS Version               : 11201
cuFFT Version                : 10300
cuRAND Version               : 10202
cuSOLVER Version             : (11, 0, 0)
cuSPARSE Version             : 11200
NVRTC Version                : (11, 1)
Thrust Version               : 100910
CUB Build Version            : 100910
cuDNN Build Version          : None
cuDNN Version                : None
NCCL Build Version           : None
NCCL Runtime Version         : None
cuTENSOR Version             : None
Device 0 Name                : GeForce GTX 1080 Ti
Device 0 Compute Capability  : 61

**Comment by awthomp** [comment]
Hi @thomasaarholt and @kmaehashi.

I've confirmed that cupy's roll can't take in a *cupy* array for the shift argument, but -- since `cp.roll` is leveraging numpy's broadcast function, this *will* work if you use a numpy array for the shift argument.

```python
import cupy as cp
import numpy as np
a = cp.arange(12).reshape(4,3)
shift = np.array([1,2,3,4])
rolled = cp.roll(a, shift)
```

Unfortunately, Thomas's suggestion to change `numpy.broadcast` to `cupy.broadcast` doesn't work as the cupy broadcast object isn't iteratable. 

We can get around this with the following code change, which basically forces a shift array to be on the CPU:

```python
# Force shift array to be numpy based on numpy broadcasting
        if isinstance(shift, cupy.ndarray):
            shift = cupy.asnumpy(shift)

        broadcasted = numpy.broadcast(shift, axis)
```

I don't think this is a good long term solution, but I plan on submitting a PR to fill this current gap.

**Comment by thomasaarholt** [comment]
That seems fine to me. I'd recommend adding a warning explaining what is happening (if a cupy array is passed) so its clear to users that the shift is being copied to the cpu.

**Comment by kmaehashi** [comment]
Closed via #4884
